man/gt-file.1
'\" t
.\"     Title: gt-file
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: 09/21/2013
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "GT\-FILE" "1" "09/21/2013" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gt-file \- Submit an array of commands listed in a file
.SH "SYNOPSIS"
.sp
.nf
\fIgt file\fR [\-\-help] [options] <file> [\-\- <qsub options>]
.fi
.SH "OPTIONS"
.PP
\-g <N>, \-g all
.RS 4
group commands into batches (N lines per group, default: N=1) use
\fI\-g all\fR
to submit all commands in a single batch
.RE
.PP
\-a <N>
.RS 4
make N attempts (resubmit up to N\-1 times if command fails)
.RE
.PP
\-t <T>
.RS 4
require h_rt=T (example: \-t 4:00:00 or \-t 14400)
.RE
.PP
\-m <M>
.RS 4
require mem_free=M (example: \-m 1G)
.RE
.PP
\-v <M>
.RS 4
require h_vmem=M (example: \-v 6G)
.RE
.PP
\-u <K>
.RS 4
add a user\-defined option
.RE
.PP
\-p
.RS 4
profile the command (report time and resource usage)
.RE
.PP
\-M
.RS 4
run MATLAB compiled code
.RE
.sp
See \(oqman qsub\(cq for qsub options\&.
.SH "DESCRIPTION"
.sp
A wrapper around \fIqsub\fR with a similar functionality as the \fIcommands2arrayjob\&.sh\fR script\&.
.sp
Each line in a file corresponds to a single task that is executed via the Bash shell\&. Hence, all these things are allowed and should work: variable substitution (e\&.g\&. $JOB_ID, $SGE_TASK_ID, etc), pipes (|), and (&&), semicolon (;), comments (#), etc\&.
.sp
Note: every line must end with a newline character (\en), in particular, the last line as well (otherwise it will not be included)\&. This is due to the way \fIwc \-l\fR counts lines\&.
.sp
The above listed options apply to the \fIgt file\fR script only\&. The script will also use the default qsub options (if any) defined in \fIgt\-setup\fR and the user config\&. These can be overriden by specifying new values after the \(em (see examples)\&.
.sp
To see all the qsub option that will be used without actually submitting a job use \fIgt \-\-dry\-run file \&...\fR\&.
.SH "EXAMPLES"
.PP
Submit the commands\&.txt file as an array job (one line one task)\&.
.RS 4
gt file commands\&.txt
.RE
.PP
Submit an array job with groups of 10 lines per task\&.
.RS 4
gt file \-g 10 commands\&.txt
.RE
.PP
Submit a job with a single task that includes all lines\&.
.RS 4
gt file \-g all commands\&.txt
.RE
.PP
Use the standard qsub options to override e\&.g\&. the inferred name\&.
.RS 4
gt file commands\&.txt \(em \-N
\fIMy commands\fR
.RE
.PP
Request 2 days of wallclock time and 8G of virtual memory per task\&.
.RS 4
gt file \-t 48:00:00 \-v 8G commands\&.txt
.RE
.PP
Submit to D2 cluster (D2blade|D2compute|D2parallel)\&.
.RS 4
gt file \-u d2 commands\&.txt
.RE
.PP
Profile each task via the \fI/usr/bin/time \-v\fR\&.
.RS 4
gt file \-p commands\&.txt
.RE
.PP
Run MATLAB compiled code (sets all the necessary environment variables)\&.
.RS 4
gt file \-M commands\&.txt Note: the MCRROOT may need to be overriden in user config and should point to the MATLAB Compiler Runtime of the appropriate version\&.
.RE
.SH "SEE ALSO"
.sp
\fBgt\fR, \fBgt st\fR, \fBgt cmd\fR
.SH "AUTHOR"
.sp
Written by Maksim Lapin\&.
.SH "REPORTING BUGS"
.sp
http://github\&.com/mlapin/gtools/issues


gt-file.sh
#!/bin/bash
#
# Submits a set of commands listed in a file
set -e
name="${GT_NAME}-file"

usage() {
  cat <<EOF
usage: ${name/-/ } [--help] [options] <file> [-- <qsub options>]

    -g <N>    group commands into batches (N lines per group, default: N=1)
              use '-g all' to submit all commands in a single batch
    -a <N>    make N attempts (resubmit up to N-1 times if command fails)
    -t <T>    require h_rt=T (example: -t 4:: or -t 14400)
    -m <M>    require mem_free=M (example: -m 1G)
    -v <M>    require h_vmem=M (example: -v 6G)
    -u <K>    add a user-defined option (available keys: ${!USER_QSUB_OPT[@]})
    -p        profile the command (report time and resource usage)
    -M        run MATLAB compiled code

See \`man qsub' for qsub options.
EOF
}

main() {
  verbose "arguments (before parsing):" "$@"

  if [[ "$1" = 'help' || "$1" = '--help' ]]; then
    show_help "file"
    exit 0
  fi

  # Parse known options
  USER_QSUB_KEY=()
  local timeit='no'
  local matlab='no'
  while getopts ":g:a:t:m:v:u:pM" opt; do
    case "${opt}" in
      g) step="${OPTARG}" ;;
      a) MAX_ATTEMPTS="${OPTARG}" ;;
      m) RES_MEMORY="${OPTARG}" ;;
      t) RES_TIME="${OPTARG}" ;;
      v) RES_VMEMORY="${OPTARG}" ;;
      u) USER_QSUB_KEY+=("${OPTARG}") ;;
      p) timeit='yes' ;;
      M) matlab='yes' ;;
      \?) echo "${name}: unknown option: -$OPTARG" >&2; usage; exit 1 ;;
    esac
  done
  shift $((${OPTIND}-1))

  if [[ $# -eq 0 ]]; then
    echo "${name}: nothing to submit."
    usage
    exit 1
  fi

  step="${step:-1}"
  cmd_file="${cmd_file:-"$1"}"
  cmd_name="${cmd_file##*/}"
  shift

  # Skip the optional '--'
  if [[ "$1" = "--" ]]; then
    shift
  fi

  # Perform certain sanity checks for the commands file
  if  [[ ! -e "${cmd_file}" ]]; then
    echo "${name}: file does not exist: ${cmd_file}" 1>&2
    exit 1
  fi

  if  [[ ! -r "${cmd_file}" ]]; then
    echo "${name}: cannot read file: ${cmd_file}" 1>&2
    exit 1
  fi

  if  [[ ! -s "${cmd_file}" ]]; then
    echo "${name}: file is empty: ${cmd_file}"
    exit 0
  fi

  if [[ "${cmd_file:0:1}" != '/' ]]; then
    cmd_file="${PWD}/${cmd_file}" # make the path absolute
  fi

  total=$(wc -l "${cmd_file}" | cut -f1 -d' ')

  if [[ "${step}" = "all" ]]; then
    step="${total}"
  fi

  # The rest (after --) are qsub options
  update_qsub_opt "$@"

  verbose "options:" \
    "attempts: ${MAX_ATTEMPTS}" \
    "file: ${cmd_file}" \
    "lines: ${total}" \
    "group by: ${step}"

  qsubmit -N "${cmd_name}" \
    -t "1-${total}:${step}" \
    "${QSUB_OPT[@]}" \
    "${LOCAL_DIR}/grun-file.sh" \
    "${LOCAL_DIR}" \
    "${MAX_ATTEMPTS}" \
    "${timeit}" \
    "${matlab}" \
    "${step}" \
    "${cmd_file}"
}

main "$@"
